contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,434 @@
+1. A `VolumeAttachment` Kubernetes API object is deleted by Kubernetes: the value for the `VolumeAttachment.metadata.deletionTimestamp` field is set.
+
+If the operation completes successfully, the external-attacher will:
+1. Remove its finalizer from the list of finalizers on the `VolumeAttachment` object permitting the delete operation to continue.
+
+If the operation fails, the external-attacher will:
+
+1. Ensure the `VolumeAttachment.Status.IsAttached` field remains true to indicate the volume is not yet detached.
+2. Set the `VolumeAttachment.Status.DetachError` field detailing the error.
+3. Create an event against the Kubernetes API associated with the `VolumeAttachment` object to inform users what went wrong.
+
+The new API object called `VolumeAttachment` will be defined as follows:
+
+```GO
+// VolumeAttachment captures the intent to attach or detach the specified
+// volume to/from the specified node.
+//
+// VolumeAttachment objects are non-namespaced.
+type VolumeAttachment struct {
+  metav1.TypeMeta `json:",inline"`
+  metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+  // Specification of the desired attach/detach volume behavior.
+  Spec VolumeAttachmentSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+  // Most recently observed status of the VolumeAttachment request.
+  // This data may not be up to date.
+  Status VolumeAttachmentStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+
+type VolumeAttachmentSpec struct {
+  // Attacher indicates the name of the volume plugin that MUST handle this
+  // request. This is the name returned by GetPluginName().
+  Attacher string `json:"attacher" protobuf:"bytes,1,opt,name=attacher"`
+
+  // Volume represents the volume that should be attached.
+  Volume AttachedVolumeSource `json:",inline" protobuf:"bytes,2,opt,name=volume"`
+
+  // NodeName indicates the node that the volume should be attached to.
+  NodeName string `json:"nodeName,omitempty" protobuf:"bytes,3,opt,name=nodeName"`
+}
+
+// VolumeAttachmentSource represents a volume that should be attached.
+// Right now only PersistenVolumes can be attached via external attacher,
+// in future we may allow also inline volumes in pods.
+// Exactly one member can be set.
+type AttachedVolumeSource struct {
+    // Name of the persistent volume to attach.
+    // +optional
+    PersistentVolumeName *string `json:"persistentVolumeName,omitempty" protobuf:"bytes,1,opt,name=persistentVolumeName"`
+
+    // Placeholder for *VolumeSource to accommodate inline volumes in pods.
+}
+
+type VolumeAttachmentStatus struct {
+  // IsAttached indicates the volume is successfully attached.
+  // This field must only be set by the entity completing the attach
+  // operation, i.e. the external-attacher.
+  IsAttached bool `json:"isAttached" protobuf:"varint,1,opt,name=isAttached"`
```

Comment:
if an attached volume becomes detached from a node for some external reason, is that expected to be reflected in this objects' status somehow? Would be worth outlining that flow, whether the external attacher is expected to continuously ensure the attached state, and what API updates would be made/observed during recovery/reattach

Reply:
CSI does not have any way how to signal that a volume has become detached. There is currently no flow to outline.

Reply:
is the external attacher expected to continuously ensure the attached state?

Reply:
Yes, external attacher continuously tries to fulfill the desired state (attached or detached). With some exponential backoff. And always reporting the latest error to `VolmeAttachment.Status.AttachError` (resp. DetachError) so internal A/D controller can pass it to pod(s) as event.